Close the cell editor when the editor's combobox is closed

imciner2 opened the pull request:
This PR makes it so the combobox grid editor is closed when the control loses focus. The loss of focus is now defined to be when the list closes, since the original KillFocus handler was never being called. This means that when the list is dismissed, the editor will close and apply the selection to the cell. Additionally, another handler needs to be added to capture the drop down event and reset the InSetFocus flag, otherwise the KillFocus handler won't function properly on GTK.

Closes [#16404](https://trac.wxwidgets.org/ticket/16404)

imciner2 committed:
Close the cell editor when the editors combobox is closed

Closes #16404

vadz commented:
Thanks, we probably should apply this to make this editor usable under macOS (AFAICS it doesn't work there at all currently, right?), but there is still a bug somewhere in port-specific code if we never get the kill focus event.

Also, the original ticket speaks only about macOS, what is the problem with GTK currently? Or does the problem only happen with these changes? If so, maybe it would be better to just do them for wxOSX only?

imciner2 commented:
The editor closing behavior actually affects all platforms from what I can tell (I have at least tested it on GTK and OSX and seen similar results). Here is what I mean:

Before changes:

![gridOriginal](https://user-images.githubusercontent.com/2262453/87146886-8d0b3080-c2a3-11ea-9a15-f171c54aab92.gif)

After changes:

![gridChanges](https://user-images.githubusercontent.com/2262453/87146900-94cad500-c2a3-11ea-95ad-dc376eaae155.gif)

In these screencasts I click outside the grid when the pointer goes into the logging area and you can see that before these changes the combobox dropdown is dismissed but the editor remained active. Also, you can see that before these changes selecting an item from the combobox list would close the dropdown but not end the editor - which is unexpected as wel.

The only parts of that ticket that I believe are platform specific are the up/down keys in the combobox list not functioning on OSX. As I mentioned in the ticket, it appears that is due to the fact we dropdown the list programatically using the `Combo()->Popup()` call on the start of editing, and that call uses a method of dropping down the list that doesn't seem to give it the correct focus. I looked around yesterday, and couldn't find a way to dropdown the list and give it focus properly (so it behaves the same as when a user clicks on the button to dropdown the list).

vadz commented:
Thanks for the GIFs, I see the problem with clicking outside the grid now. However it doesn't exist in wxMSW, even though it clearly exists in wxGTK. I'd like to understand why...

But I think you're right that people expect the editor to be dismissed after selecting an item from the dropdown, and this doesn't happen under MSW neither right now. So doing it from `wxEVT_COMBOBOX_CLOSEUP` handler looks like a good idea, I just wonder if we can do it in some less clumsy way than by generating an artificial focus loss event.

imciner2 commented:
> I just wonder if we can do it in some less clumsy way than by generating an artificial focus loss event.

I thought about that, and one way would be to generate the `wxEVT_GRID_HIDE_EDITOR` directly in the `wxEVT_COMBOBOX_CLOSEUP` handler. Doing that requires having a reference/pointer to the grid the editor is in though, and I don't think the editor currently stores that. It might be possible to expose that as a public function inside `wxGridCellEditorEvtHandler` so then it can get the grid from the event handler. One other thing was the large TODO comment that is located at [`grideditors.cpp:90`](https://github.com/wxWidgets/wxWidgets/blob/master/src/generic/grideditors.cpp#L90), which made me leery of duplicating that code into another place in the codebase. 

Removing the dummy focus event generation and directly closing the editor from the `wxEVT_COMBOBOX_CLOSEUP` would also remove the need for the `wxEVT_COMBOBOX_DROPDOWN` event handler - who's only purpose is to reset the flag preventing the kill focus handler from killing the editor when it is being opened so that it can be killed when the popup closes.

imciner2 commented:
@vadz what are your thoughts on that VC6 TODO statement? Is VC6 a supported compiler anymore? If it isn't, then I can probably fix that part up and it would clean up this implementation as well.

vadz commented:
Sorry, didn't have time to look at this yet, but VC6 is definitely not supported since 3.1.0, see [wx/compiler.h](https://github.com/wxWidgets/wxWidgets/blob/b7057c336f3b9fc4041171b107dd1f55404a65d3/include/wx/compiler.h#L39).

imciner2 commented:
Thanks, I did a quick glance through the code and am somewhat confused by the uses of `CallAfter` right now. Looking in the codebase, it appears that its definition is guarded by a test for VC8+, so it's not defined when building on VC7:
```
// Currently VC7 is known to not be able to compile CallAfter() code, so
// disable it for it (FIXME-VC7).
#if !defined(__VISUALC__) || wxCHECK_VISUALC_VERSION(8)
    #include "wx/meta/removeref.h"

    #define wxHAS_CALL_AFTER
#endif
```
and then the actual definitions are guarded by `wxHAS_CALL_AFTER` (https://github.com/wxWidgets/wxWidgets/blob/master/include/wx/event.h#L3660).

However, it appears that there are usages in the MSW code that are doing unguarded calls to it (see https://github.com/wxWidgets/wxWidgets/blob/master/src/msw/popupwin.cpp#L244, https://github.com/wxWidgets/wxWidgets/blob/master/src/common/combocmn.cpp#L597, and others).

Are these unguarded calls a mistake? Or does `CallAfter` actually work on VC7 and the check is wrong?

vadz commented:
I think it means that the build is broken with VC7 and nobody noticed... Which looks like a good reason to drop support for it, if it can facilitate things.

BTW, grepping for them now, there are checks for VC5 (!!) in `wx/defs.h` that should be definitely removed.

imciner2 commented:
In that case, I think it is worth dropping VC7 (it's from 2002 I think?). Trying to support it would mean having to fix a few different places in the code that use `CallAfter`, and I think since no one noticed issues with it (and some of those places have been in the code for a few years), it is probably not widely used with 3.1. It would certainly make things nicer if I could cleanup the calls to not have to propagate that event. Although, can we really remove the `wxEVT_GRID_HIDE_EDITOR` event without breaking user code? It looks like it isn't in the documented API, but the handler for it is `protected` not `private`, so people could theoretically be tapping onto the event still.

vadz commented:
VC7 is MSVS 2003, but one year hardly makes a difference and I agree that we can drop support for it. The thing/problem is that I wanted to also drop support for MSVS 2005 and 2008 and hence the old `.vcproj` project files at once too, and I'm not sure if there are no people out there using at least the latter (I worked on a project which used VC8, a.k.a. MSVS 2005, up until less than a year ago!). So we'll probably still keep them for 3.2 -- but, again, let's drop VC7. Of course, you don't have to do this, i.e. you can just use `CallAfter()` as you see fit, and I'll remove the tests for it later (although if you make a PR for this too, I certainly wouldn't object at all).

Code search for `wxEVT_GRID_HIDE_EDITOR` on GitHub and Debian code search doesn't find any matches for it in the user code, so we probably could remove it, but OTOH I think it can be useful to know about the editor being dismissed, so maybe it would be better to generate it?

imciner2 commented:
If people want to know that the editor is being dismissed, then we already export the event `wxEVT_GRID_EDITOR_HIDDEN`, which is generated inside `wxGrid::DoDisableCellEditControl()`. There is also already a public `wxGrid::HideCellEditControl()` that can be used to close the current editor (which will actually cause the kill-focus handler of the editor to be called), so I don't really think we need to have an event that allows the editor to be hidden.

vadz commented:
Sorry, I confused `wxEVT_GRID_EDITOR_HIDDEN` and `wxEVT_GRID_HIDE_EDITOR`. I am definitely fine with removing the latter.

vadz commented:
Oh, I knew I was forgetting something: in fact, I have this [6 year (!) old commit](https://github.com/vadz/wxWidgets/commit/09b0723d9a36dbddd519c9ed6996e56bdff8adce) which already does this. It probably doesn't apply any more, but it clearly shows that I planned to do it myself, and just never finished it...

imciner2 committed:
Replace wxEVT_GRID_HIDE_EDITOR usage with CallAfter

This simplifies the code by removing the need for a special event,
and also means the combox popup handler is no longer needed to reset
the inSetFocus flag for the closing handler.

imciner2 commented:
Thanks! That was basically what I was thinking it would look like. I did have to introduce a helper `wxGridCellEditorEvtHandler::Grid()` method so that I could access the grid from the actual editors (so `CallAfter()` could be used in the datetime and combobox editors). There was also one direct call to the `DisableCellEditControl()` function inside the event handler's `OnKeyDown()` handler that I changed to use `CallAfter()` now, and I am surprised that usage hasn't caused issues before (since the comments mentioned that the reason for the event/`CallAfter` was to prevent crashes due to the direct call).

I think this takes care of everything for the grid parts. You can do the remaining pieces to fully deprecate VC6/VC7 from the codebase (and cleanup those VC5 instances also I guess) in an additional PR as you get time.

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -1679,6 +1676,19 @@ wxString wxGridCellChoiceEditor::GetValue() const
 
+void wxGridCellChoiceEditor::onComboCloseUp(wxCommandEvent& WXUNUSED(evt))
+{
+    wxGridCellEditorEvtHandler* evtHandler = wxDynamicCast(m_control->GetEventHandler(),
```
Another thing: could we avoid this repeated call to dynamic cast if we just caught `CLOSEUP` event in `evtHandler` itself? It should live for long enough, AFAICS, and this could make things simpler...

imciner2 replied:
I did it this way to avoid giving the event handler knowledge about the editors themselves. Also, I think it is more visible if it is associated with the editor: this way looking at the editor code it is obvious the combobox event is being handled.

vadz replied:
This is a good point, thanks.

vadz reviewed (commented):
Sorry, I'm still not completely convinced this is the best way to do this. I haven't had time to actually test this myself yet, so sorry in advance if I'm missing something obvious here...

imciner2 commented on src/generic/grideditors.cpp:
```diff
@@ -100,7 +93,7 @@ void wxGridCellEditorEvtHandler::OnKeyDown(wxKeyEvent& event)
         case WXK_ESCAPE:
             m_editor->Reset();
-            m_grid->DisableCellEditControl();
+            m_grid->CallAfter(&wxGrid::DisableCellEditControl);
```
Two reasons: consistency with the other calls and the fact that the comment in the other event handler says that directly calling this function is not good because when the editor is disabled the event handler could be destroyed - leading to invalid memory access (that was why the kill focus handler used the event to request the closing of the editor). This call is also inside an event handler, so it would run into the same issue of the event handler object being destroyed before this event is finished processing. The case where a crash like this could happen is probably rare, since it would require pointers to be overwritten in a very short timespan, but since we already know how to fix it, I think it is better to fix it before someone reports a crash with it (or has to try and track down a crash in it).

vadz replied:
Sometimes postponing the event can result in unexpected problems too, e.g. if a handler for some event running immediately after this `wxEVT_KEY_DOWN` relies on the editor being already dismissed, but I guess it's indeed better for consistency, so let's do it like this and see if anybody complains...

imciner2 commented on include/wx/generic/grideditors.h:
```diff
@@ -27,6 +27,8 @@ class wxGridCellEditorEvtHandler : public wxEvtHandler
     {
     }
 
+    wxGrid* Grid() { return m_grid; }
```
Yes, this would be a cleaner thing to do.

imciner2 replied:
I've gone ahead and done this and pushed the squashed commits.

imciner2 commented on include/wx/generic/grideditors.h:
```diff
@@ -335,6 +337,8 @@ class WXDLLIMPEXP_ADV wxGridCellChoiceEditor : public wxGridCellEditor
 protected:
     wxComboBox *Combo() const { return (wxComboBox *)m_control; }
 
+    void onComboCloseUp(wxCommandEvent& evt);
```
Fixed.

vadz commented:
Thanks again for improving this!
